### PR TITLE
[XCOFF][obj2yaml] support parsing auxiliary symbols for XCOFF

### PR DESCRIPTION
**Pull request by EsmeYi**
This PR adds the support for parsing auxiliary symbols of XCOFF object file for obj2yaml.

### CODE REVIEW

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:116]: Could these be passed by reference, since they (presumably) can't be null? Same below.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:165]: Nit: it might be slightly easier to read if there was a blank line before this `if`.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:210]: 1) `NumOfAuxSym` is a `uint8_t`, so `I` should be too.
2) `++I` (https://llvm.org/docs/CodingStandards.html#prefer-preincrement)

Also, this bit of code is pretty substantial. Let's pull this case body into a separate function for ease of following the `switch`.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:238]: Same comments as above.

**Review by jh7370 — COMMENTED** [review]
I've got nothing significant to add: one of the more XCOFF-knowledgeable developers needs to review this too.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:135]: `dumpFuncAuxSym` would be clearer, I think? I'd rename the variables and other similar functions similarly.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:216]: I think you can just do `std::vector<std::unique_ptr<XCOFFYAML::AuxSymbolEnt>> &AuxEntTbl = Sym.AuxEntries;`. Then you don't need to move at the end of the block.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:174]: Ping?

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:219]: This loop is still inline in the switch when I asked for it to be pulled out into its own function. Please address all my comments or explicitly say you're either not doing them (and why not) or that you're going to do them in a different diff update. See https://llvm.org/docs/CodeReview.html#acknowledge-all-reviewer-feedback.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:244]: Same as above - please move this into a function, as requested before.

File: llvm/lib/ObjectYAML/XCOFFYAML.cpp
Comment (diggerlin) [inline llvm/lib/ObjectYAML/XCOFFYAML.cpp:298]: suggestion add a lambda helper function

```
 auto ResetAuxSym =
      [&](auto* AuxEnt) {
        if (!IO.outputting())
          AuxSym.reset(AuxEnt);
      };
```

and then change the code 
```
 if (!IO.outputting())
      AuxSym.reset(new XCOFFYAML::ExcpetionAuxEnt()); 
```
  to 
  
 ` ResetAuxSym (new XCOFFYAML::ExcpetionAuxEnt());`

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:135]: Nice.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp]: Thank you for your suggestions! (Sorry, I didn't notice this until after the update).
Although I have no objection about your proposal, actually I'm not sure which implementation is more preferable since current style aligns with other formats like COFF or ELF. And does it make more sense to put the lengthy implementation type code here instead of in the header file?
Hi James @jh7370 , what do you think?

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp]: I didn't fully catch your comments before and sorry about it. Now the whole case body is pulled out into a separate function instead of partial codes.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (jh7370) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp]: I think the implementation of the constructors can be in the cpp, should you choose to go down that route. I think having constructors will likely make things cleaner, but you lose the POD nature of the objects. I've got no strong feeling on whether that is an issue or not, and leave it up to you two and other XCOFF stakeholders to agree on that (I don't think it needs to match COFF/ELF etc specifically).

**Review by jh7370 — COMMENTED** [review]
@EsmeYi, please don't resolve comments that I've made, as I need to look back through them to check that they've been addressed as I expected. See https://discourse.llvm.org/t/rfc-github-pr-resolve-conversation-button/73178.

Nothing new to add at this point. I'll take a further look through once Digger is basically happy.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (diggerlin) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp]: The functionality of the function is to create a new object `SectAuxEntForStat` and initiate with `XCOFFSectAuxEntForStat` , using constructor is object oriented.
it is a suggestion , feel free not to modify if you want to keep current implement.

Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp]: Thank you @diggerlin @jh7370 , well then I would prefer to keep current style.

**Review by jh7370 — APPROVED** [review]
LGTM, once @diggerlin is happy.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (diggerlin) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:281]: ditto for assert.

File: llvm/test/tools/obj2yaml/XCOFF/aix.yaml
Comment (diggerlin) [inline llvm/test/tools/obj2yaml/XCOFF/aix.yaml:55]: just curiosity, you have   `AUX_CSECT` test in aux0symbols,yaml, why you add test case here ?

Comment (EsmeYi) [reply llvm/test/tools/obj2yaml/XCOFF/aix.yaml:55]: NumberOfAuxEntries was set to 1 but non auxiliary symbol is defined in this case, which was allowed before but will be illegal after this PR. And `aix.yaml` is a basic test for general behavior of xcoff2yaml and I think it would be better to add such test instead of remove the set of `NumberOfAuxEntries: 1` here.

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (diggerlin) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:152]: Suggest that we  provide more detail info for invalid Symbol (for example symbol index), let user can locate the specific invalid symbol.    

and some test case for the Error scenarion.

File: llvm/lib/ObjectYAML/XCOFFYAML.cpp
Comment (diggerlin) [reply llvm/lib/ObjectYAML/XCOFFYAML.cpp:304]: the error print out, but the it do not return function immediately. it go to the  `ResetAuxSym(new XCOFFYAML::ExcpetionAuxEnt())` ,  since the ExcpetionAuxEnt is not support in the 32bit, do not need to run the `ResetAuxSym(new XCOFFYAML::ExcpetionAuxEnt());' is it correct ?

File: llvm/test/tools/obj2yaml/XCOFF/aix.yaml
Comment (diggerlin) [reply llvm/test/tools/obj2yaml/XCOFF/aix.yaml:55]: thanks for explain.

File: llvm/lib/ObjectYAML/XCOFFYAML.cpp
Comment (EsmeYi) [reply llvm/lib/ObjectYAML/XCOFFYAML.cpp:304]: OK, I see. Thanks!

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (diggerlin) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:253]: Sym.SymbolName does not include the storage mapping class; several symbols may have the same name. I prefer to use symbol index here, but if we use symbol index, a new data member 'SymbolIndex' needs to be added to 'XCOFFYAML::Symbol.' I am not sure whether it is worth it for us to do so or not.

Comment (jh7370) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:253]: If I'm not mistaken, the calling code can know the index by changing the loop it is using. The index could then be passed into this function without changing `Symbol`.

Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:258]: The index can be easily caught by `uint32_t SymbolIndex = Obj.getSymbolIndex(SymbolEntRef.getEntryAddress());`
I just pushed the updated commit. Thanks! @diggerlin and @jh7370

File: llvm/tools/obj2yaml/xcoff2yaml.cpp
Comment (diggerlin) [inline llvm/tools/obj2yaml/xcoff2yaml.cpp:143]: for all the 64-bits auxiliary symbol entries, there is a field named `x_auxtype` , I think  we need to print it out too, although the field can derived from the content of  the auxiliary by user.    obj2yaml is dump tool , we need to display all the fields in the auxiliary symbol entries

Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:143]: If I understand you correctly, I think currently we have print the aux type in the yaml file regardless of whether it is 32-bit or 64-bit. There is a constructor in each aux symbol struct, like `FileAuxEnt():AuxSymbolEnt(AuxSymbolType::AUX_FILE) {}`, and we can verify this in the test file llvm/test/tools/obj2yaml/XCOFF/aux-symbols.yaml

( Although auxiliary symbol entries in XCOFF32 do not really have the x_auxtype field, it's no harm to print the info in yaml and it's helpful when transforming yaml to object.

Comment (diggerlin) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:143]: > If I understand you correctly, I think currently we have print the aux type in the yaml file regardless of whether it is 32-bit or 64-bit. There is a constructor in each aux symbol struct, like `FileAuxEnt():AuxSymbolEnt(AuxSymbolType::AUX_FILE) {}`, and we can verify this in the test file llvm/test/tools/obj2yaml/XCOFF/aux-symbols.yaml
> 
> ( Although auxiliary symbol entries in XCOFF32 do not really have the x_auxtype field, it's no harm to print the info in yaml and it's helpful when transforming yaml to object.

My understand we should print out the fields as it is in the object file.  what is James' opinion ? @jh7370

Comment (jh7370) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:143]: Sorry, I'm not really sure I follow the conversation, as it is too much in the file format details. Some general comments:
1) We need enough info in the YAML produced by obj2yaml for yaml2obj to be able to create an identical object as was passed to obj2yaml.
2) This means we can potentially omit fields from the YAML dump if they have a value matching the default value yaml2obj would use if the field is omitted.
3) The less data we have in the YAML, the better, as long as the first point is conformed to. This is because it'll be easier to read, and if somebody is using obj2yaml to generate the YAML that will be used as the input in a test case, it is minimal and has no redundant information.

Comment (EsmeYi) [reply llvm/tools/obj2yaml/xcoff2yaml.cpp:143]: Thanks!
The aux type is not in the XCOFF32 object, but it is needed in YAML to help yaml2obj create the identical object.

**Review by diggerlin — APPROVED** [review]
LGTM

File: llvm/lib/ObjectYAML/XCOFFYAML.cpp
Comment (bzEq) [inline llvm/lib/ObjectYAML/XCOFFYAML.cpp:288]: What if `IO.outputting()` is `true`, will `AuxEnt` be deleted?

**Pull request merged**
